web/shared-beans/src/main/java/io/deephaven/web/shared/data/RangeSet.java

```diff
@@ -37,210 +39,362 @@ public static RangeSet ofItems(long... items) {
+    }
+
+    private static class RangeAccumulator {
+        private final List<Range> replacement = new ArrayList<>();
+
+        public void appendRange(Range range) {
+            if (!replacement.isEmpty()) {
+                Range lastSeen = replacement.get(replacement.size() - 1);
+                Range overlap = lastSeen.overlap(range);
                 if (overlap != null) {
-                    // replace the range that we are merging, and start the slice here instead
-                    merged = overlap;
-                    proposedIndex--;
-                    // TODO this will make the loop start here, considering this item twice. not ideal, but not a big
-                    // deal either
+                    replacement.set(replacement.size() - 1, overlap);
+                } else {
+                    replacement.add(range);
                 }
+            } else {
+                replacement.add(range);
             }
-            // "end" represents the last item that needs to be merged in to the newly added item. if no items are to be
-            // merged in, then end will be proposedIndex-1, meaning nothing gets merged in, and the array will grow
-            // instead of shrinking.
-            // if we never find an item we cannot merge with, the end of the replaced range is the last item of the old
-            // array, which could result in the new array having as little as only 1 item
-            int end = sortedRanges.length - 1;
-            // until we quit finding matches, test subsequent items
-            for (int i = proposedIndex; i < sortedRanges.length; i++) {
-                Range existing = sortedRanges[i];
-                Range overlap = existing.overlap(merged);
-                if (overlap == null) {
-                    // stop at the item before this one
-                    end = i - 1;
-                    break;
-                }
-                merged = overlap;
-            }
-            int newLength = sortedRanges.length - (end - proposedIndex);
-            assert newLength > 0 && newLength <= sortedRanges.length + 1;
-            Range[] newArray = new Range[newLength];
-            if (proposedIndex > 0) {
-                System.arraycopy(sortedRanges, 0, newArray, 0, proposedIndex);
-            }
-            newArray[proposedIndex] = merged;
-            if (end < sortedRanges.length - 1) {
-                System.arraycopy(sortedRanges, end + 1, newArray, proposedIndex + 1, sortedRanges.length - (end + 1));
-            }
-            sortedRanges = newArray;
+            // assert isSorted();
+        }
+
+        public void appendRanges(List<Range> ranges) {
+            appendRange(ranges.get(0));
+            replacement.addAll(ranges.subList(0, ranges.size() - 1));
+            // assert isSorted();
+        }
+
+        public void appendRanges(List<Range> ranges, long firstItemSubindex) {
```

Comment:
That this new method is unused is one of my qualms about this change - or at least is an opportunity to improve. A binary search when we get started on add/remove/shift, bulk adding as we encounter consecutive untouched ranges, or getting the index of the last range touched so we can bulk add the rest would let us skip ~O(n) long `+` ops or Range.overlap (i.e. two compares) for each range. Not going to make a huge difference, compared to the work so far, but could buy us a little more breathing room.